src/arch/io/network.cc

```diff
@@ -621,6 +621,430 @@ void linux_tcp_conn_t::on_event(int /* events */) {
+                probably signalled by on_shutdown(). */
+                throw tcp_conn_read_closed_exc_t();
+            }
+            /* Go around the loop and try to read again */
+            continue;
+        case SSL_ERROR_WANT_WRITE:
+            /* Though we are reading, a TLS renegatiotain may occur at any time
+            requiring a write. Wait for a notification from the event queue, or
+            for an order to shut down */
+            {
+                linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_out);
+                wait_any_t waiter(&watch, &closed);
+                waiter.wait_lazily_unordered();
+            }
+            if (closed.is_pulsed()) {
+                /* We were closed for whatever reason. Whatever signalled
+                us has already called on_shutdown(). */
+                throw tcp_conn_read_closed_exc_t();
+            }
+            /* Go around the loop and try to write again */
+            continue;
+        case SSL_ERROR_SYSCALL:
+            // An error with the underlying IO.
+            io_err = ERR_get_error();
+            if (io_err != 0) {
+                // The Openssl error queue has data on the error.
+                on_shutdown();
+                return;
+            }
+
+            // if ret is 0 that indicates an EOF in violation of the protocol.
+            if (0 == ret) {
+                on_shutdown();
+                return;
+            }
+
+            // Otherwise consult errno.
+            switch (get_errno()) {
+            case EPIPE: case EHOSTUNREACH: case ENETDOWN:
+            case EHOSTDOWN: case ECONNRESET: case ENOTCONN:
+                /* These errors are expected to happen at some point in practice */
+                on_shutdown();
+                return;
+            default:
+                /* Unknown error. This is not expected, but it will probably happen sometime so we
+                shouldn't crash. */
+                on_shutdown();
+                return;
+            }
+        default:
+            // Some other error.
+            on_shutdown();
+            return;
+        }
+    }
+}
+
+void linux_secure_tcp_conn_t::shutdown() {
+    assert_thread();
+
+    int ret;
+
+    // If we have not already received a "close notify" alert from the peer,
+    // then we should send one. If we have received one, this loop will respond
+    // with our own "close notify" alert.
+    while (true) {
+        ERR_clear_error();
+
+        if ((ret = SSL_shutdown(conn)) >= 0) {
+            break; // "close notify" was either sent or received.
+        }
+
+        // A negative ret indicates as error.
+        int err_code = SSL_get_error(conn, ret);
+        if (SSL_ERROR_WANT_READ == err_code) {
+            linux_event_watcher_t::watch_t watch(get_event_watcher(), poll_event_in);
+            wait_any_t waiter(&watch);
+            waiter.wait_lazily_unordered();
```

Comment:
If it's not included yet, you'll need to `#include "arch/timing.hpp"` to get `signal_timer_t`.